Checkup config: read and parse checkup's config from ConfigMap object

orelmisan opened the pull request:
This add the ability to read and parse checkup's configuration from a ConfigMap object.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/66bff36aa0f3db018bcfc23d180186e755ab5bbb..edaaa90f089d16eb2f227ae7bdefa8ee0cba5bb1): Fixed a problem discovered by the linter.

EdDev commented on kiagnose/mainflow.go:
```diff
@@ -20,12 +20,43 @@
 	return l.Run()
 }
+
+func readConfigMapFullNameFromEnv() (namespace, name string, err error) {
```
This should not be here, but hidden in some package.

orelmisan replied:
Done.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/81957fe3100b8474c3dd428df0bb0f9be635f4e7..7f40b7036a5466a7bec415ee999f346c897e091b): Changed a commit message.

orelmisan commented:
Changes: Optimized the unit test code.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/65691c0adf658ea9aa0c19573d97431a00913552..5ba741e6a9ce2583af23c85a1141141f72f73022): Fixed linter problems.

EdDev commented on kiagnose/internal/checkup/spec_test.go:
```diff
@@ -0,0 +1,218 @@
+	const missingEnvVarErr = "missing required environment variable(s)"
+
+	t.Run("run successfully when required environment variables exists", func(t *testing.T) {
+		os.Setenv(checkup.ConfigMapNamespaceEnvVarName, configMapNamespace)
+		defer os.Unsetenv(checkup.ConfigMapNamespaceEnvVarName)
+		os.Setenv(checkup.ConfigMapNameEnvVarName, configMapName)
+		defer os.Unsetenv(checkup.ConfigMapNameEnvVarName)
```
I prefer our unit tests to be "pure", in the sense they are not to depend on the OS.

Per what I actually see in these tests, it is also not needed. The production code does not call `os.*`, so I am unclear why it is used at all.

orelmisan replied:
This was removed.

EdDev commented on kiagnose/internal/checkup/spec_test.go:
```diff
@@ -0,0 +1,218 @@
+	assert.NotNil(t, specLoader)
+}
+
+func checkNewSpecLoaderFailure(t *testing.T, expectedError string) {
```
I find this helper problematic for several reasons:
- It hides the main test operation, mixing both execution of the tested production code and the assertions.
- That you expect a specific error text, which does not seem to be a public API (i.e. it can change and therefore break all these tests).
- That you use `os.Environ()` and therefore depend on the OS, when there is nothing that requires it.
- That you create a fake client, but you do not really have to.

EdDev commented on kiagnose/internal/checkup/spec_test.go:
```diff
@@ -0,0 +1,218 @@
+
+	specLoader, err := checkup.NewSpecLoader(fakeClient, env)
+	assert.ErrorContains(t, err, expectedError)
+	assert.Nil(t, specLoader)
```
It is a common practice to ignore the returned value in case of error.
No one should trust the value in case of an error, so we do not need to check it.

EdDev commented on kiagnose/mainflow.go:
```diff
@@ -20,12 +20,50 @@
+	"log"
+	"os"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
```
I prefer not to see such a dependency at this top level.
Barring it deep under an inner package makes it cleaner IMO. (e.g. [this](https://github.com/kiagnose/kiagnose/pull/9/files#diff-c810fb751f76dc7919d84b25d9611683296c4e3648b45d588ba7db3af1954038R20))

orelmisan replied:
Done.

EdDev commented on kiagnose/mainflow.go:
```diff
@@ -20,12 +20,50 @@
+		return err
+	}
+
+	env := os.Environ()
```
This is an implementation detail I prefer to be deep inside the package that implements it.

orelmisan replied:
This was removed.

EdDev commented on kiagnose/mainflow.go:
```diff
@@ -20,12 +20,50 @@
+	}
+
+	env := os.Environ()
+	specLoader, err := checkup.NewSpecLoader(clientset, env)
+	if err != nil {
+		return err
+	}
+
+	checkupSpec, err := specLoader.Load()
+	if err != nil {
+		return err
+	}
```
I would have preferred to see a separation between the `checkup` package and the loading/importing of the configuration.
To have such a separation, I imagine a totally independent package that is instantiated and then instructed to load the configuration, similar to what you just did.

I am unclear why I need a `*Loader`, I think from a usage perspective, this can be enough:
```
conf, err := config.New(clientset)
if err != nil { return err }

confData, err := conf.Load()
if err != nil { return err }
```

In the `config` package I expect to see a `config` file that has the general/common parts and in a `configmap` the things specific to configmaps. Same for `envvar` and other implementation specifics.
Considering that we do want to more in the future away from them, we want to change as little as possible at the business logic or common areas.

orelmisan commented:
Changes: Fixed code review comments + linter errors.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/baae1eaeddbcede3194a2bbc676c2c7509cec323..fd48f7f2987eb4a710e672574b78f36e29621257): Moved types form loader_test.go to outside of Test*() functions.
Also changed their field names according to an offline code-review comment by @ormergi.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/fd48f7f2987eb4a710e672574b78f36e29621257..cf1d4c7071850ada0d22df9da2b04a5fa79df61b): Changed Config.Params to Config.EnvVars of type `[]corev1.EnvVar`.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/e0d7a0f6299fb4fd640e46192b4f7b0bc3210846..bcb79833db1d963b3cc7de46893568817dba6841): Added `config.ErrRoleNameIsIllegal`.

EdDev commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,211 @@
+	return strings.Split(trimmedString, "\n")
+}
+
+func splitRoleName(fullName string) (namespace, name string, err error) {
```
Why are you using named return values? This is not a common practice and is usually reserved for cases where you must (e.g. using a defer that wants to change the returned value).

orelmisan replied:
Our linter requires it:
```
./automation/make.sh 
kiagnose/internal/config/loader.go:179:1: unnamedResult: consider giving a name to these results (gocritic)
func splitRoleName(fullName string) (string, string, error) {
^
```

EdDev commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,211 @@
+)
+
+type Loader struct {
+	client             kubernetes.Interface
+	configMapNamespace string
+	configMapName      string
```
Looking at this after looking at all the rest of the code, it looks to me like these 3 values are the only ones needed for a loader. All the rest are results of intermediate steps which do not really need to be preserved in this object state.

orelmisan replied:
Done.

EdDev commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,284 @@
+	config.ConfigMapNameEnvVarName:      configMapName,
+}
+
+var (
+	clusterRoleNamesList = []string{"cluster_role1", "cluster_role2"}
+	expectedClusterRoles = []*rbacv1.ClusterRole{
+		{
+			TypeMeta:   metav1.TypeMeta{Kind: "ClusterRole", APIVersion: rbacv1.GroupName},
+			ObjectMeta: metav1.ObjectMeta{Name: "cluster_role1"},
+		},
+		{
+			TypeMeta:   metav1.TypeMeta{Kind: "ClusterRole", APIVersion: rbacv1.GroupName},
+			ObjectMeta: metav1.ObjectMeta{Name: "cluster_role2"},
+		},
+	}
+
+	roleNamesList = []string{"default/role1", "default/role2"}
+	expectedRoles = []*rbacv1.Role{
+		{
+			TypeMeta:   metav1.TypeMeta{Kind: "Role", APIVersion: rbacv1.GroupName},
+			ObjectMeta: metav1.ObjectMeta{Name: "role1", Namespace: "default"},
+		},
+		{
+			TypeMeta:   metav1.TypeMeta{Kind: "Role", APIVersion: rbacv1.GroupName},
+			ObjectMeta: metav1.ObjectMeta{Name: "role2", Namespace: "default"},
+		},
+	}
+)
```
It seems to me like all globals here are used in specific tests (or groups) and not shared between many.
So it would be safer and more easily read if placed near the usage as possible.

Please try to follow this with other variables, I have may missed others.

EdDev commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,284 @@
+		fakeClient := fake.NewSimpleClientset()
+		loader, err := config.NewLoader(fakeClient, goodEnv)
+		assert.NoError(t, err)
+		assert.NotNil(t, loader)
```
I think that checking that it is not a nil is not very useful, could you check against something you expect to see in the output that has some explicit meaning?
Such asserts have value during development, but less value at the end (unless you can check for some explicit result).

EdDev commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,284 @@
+		},
+	}
+
+	fakeClient := fake.NewSimpleClientset()
```
You can inline this, but is it even needed for these tests?

orelmisan replied:
Replaced it with nil.

EdDev commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,284 @@
+
+func TestNewConfigLoaderShould(t *testing.T) {
+	t.Run("run successfully when required environment variables exists", func(t *testing.T) {
+		fakeClient := fake.NewSimpleClientset()
```
Can you pass a `nil` instead of a fake? You do not really need the fake in these tests.

orelmisan replied:
Replaced it with nil.

EdDev commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,284 @@
+	}
+}
+
+func newFakeClientset(namespace,
```
This seems to be a clientset with stuff in it. Can you please name it to express that?

EdDev commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,284 @@
+	}
+}
+
+func newExpectedConfig(image, timeout string,
+	envVars []corev1.EnvVar,
+	clusterRules []*rbacv1.ClusterRole,
+	roles []*rbacv1.Role) *config.Config {
```
I would prefer one liner, even if it is long. It is hard to read it like this.

Same for the others.

orelmisan commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,211 @@
+	return l.checkupConfig, nil
+}
+
+func (l *Loader) getConfigMapFullNameFromEnv(env map[string]string) (err error) {
```
1. The named returned value was a leftover, thanks for catching it.
2. Agreed.

orelmisan replied:
Done.

orelmisan commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,211 @@
+func NewLoader(client kubernetes.Interface, env map[string]string) (*Loader, error) {
+	l := &Loader{
+		client:        client,
+		checkupConfig: &Config{},
```
This was changed, and taken out of config.Loader.

orelmisan commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,211 @@
+
+	if data[RolesKey] != "" {
+		l.roleNames = parseListSeparatedByNewlines(data[RolesKey])
+	}
```
Done.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/bcb79833db1d963b3cc7de46893568817dba6841..ea90db57115a224a2328eae590b1f5684e32ef9a): Fixed comments from code review in production code.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/ea90db57115a224a2328eae590b1f5684e32ef9a..cbe4d70fbee84e205150a79cc1ee17385fbf6fc8): Fixed comments from code review in unit test code.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/cbe4d70fbee84e205150a79cc1ee17385fbf6fc8..3df113b0ca02f8742dc59b9b13d37d2932a2f8e8): Added a test case - fail when ConfigMap object doesn't exists.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/3bafbf849f1e0ab20ce67258d9c9e6ad338650d1..f9bdef8ead13148dffaecf433eb81abe87bd5590): Moved low level K8s client usage to a dedicated function.
Try to get ClusterRole and Role objects, only after successful parsing of their names.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/f9bdef8ead13148dffaecf433eb81abe87bd5590..68792f2fa76a70b2ddcd45295ec313b30c4d49f9): Extracted the parsing logic from config.Loader to a new entity.

orelmisan commented:
Change: Fixed code review comments regarding unit test code.

EdDev commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,141 @@
+	return clusterRoles, nil
+}
+
+func (l *Loader) getRoles(roleNames []string) ([]*rbacv1.Role, error) {
```
nit: You do not need the get really, it is redundant.

EdDev commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,141 @@
+	var roles []*rbacv1.Role
+
+	for _, roleName := range roleNames {
+		splitName, err := splitRoleName(roleName)
```
nit: `fullRoleName` ?

EdDev commented on kiagnose/internal/config/parser.go:
```diff
@@ -0,0 +1,187 @@
+	roleNames          []string
+}
+
+func newConfigMapParser(client kubernetes.Interface, env map[string]string) (*configMapParser, error) {
```
Can you also name the file with relation to the ConfigMap? e.g. `cmparser`

EdDev commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,141 @@
+}
+
+func (l *Loader) Load() (*Config, error) {
+	if err := l.parser.Get(); err != nil {
+		return nil, err
+	}
```
It does not seem to be a responsibility of a parser to get the object. A parser parses, it is not suppose to know how to fetch the object.

But I think this can be improved in parallel to the next step, so for now I tend to just let it roll.

orelmisan replied:
I agree that its not the parser's responsibility.

EdDev commented on kiagnose/internal/config/parser.go:
```diff
@@ -0,0 +1,187 @@
+func (cmp *configMapParser) Get() error {
+	configMap, err := cmp.client.CoreV1().ConfigMaps(cmp.configMapNamespace).Get(context.Background(), cmp.configMapName, metav1.GetOptions{})
+	if err != nil {
+		return err
```
I think you are not testing this path.

orelmisan replied:
It is tested in a scenario when the ConfigMap doesn't exists.
When we'll have E2E tests, they should also test read permission problems.

EdDev replied:
The unit tests do not simulate the error occurring here, which means this condition is not evaluated and covered.

EdDev commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,141 @@
+
+		role, err := l.client.RbacV1().Roles(splitName.Namespace).Get(context.Background(), splitName.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
```
I think you are not testing this path.

orelmisan replied:
It is tested in a scenario when the Role doesn't exists.
When we'll have E2E tests, they should also test read permission problems.

orelmisan replied:
The call to `Get()` returns an error if there is no Role with the requested `namspace/name`.

EdDev commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,141 @@
+	for _, name := range clusterRoleNames {
+		clusterRole, err := l.client.RbacV1().ClusterRoles().Get(context.Background(), name, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
```
I think you are not testing this path.

orelmisan replied:
It is tested in a scenario when the ClusterRole doesn't exists.
When we'll have E2E tests, they should also test read permission problems.

EdDev replied:
This still needs to be covered by unit tests.

orelmisan replied:
The call to `Get()` returns an error if there is no ClusterRole with the requested name.

EdDev reviewed (commented):
I had some comments but I am ok with proceeding in this state as is and fixing/improving as a followup PR.

Review and check if you want to take some of the suggestions now.
And most importantly, please check this is working ok against a real `ConfigMap` if you can.

orelmisan commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,220 @@
+	return checkupConfig, nil
+}
+
+func (l *Loader) getClusterRoles(clusterRoleNames []string) ([]*rbacv1.ClusterRole, error) {
```
`l.clusterRoles` looks off to me.

EdDev replied:
I do not like `get`, but this is a nit, I will not insist.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/0931821973a9544544444c33b6c711e5f2bb97ec..fc259ce518132787efff1cf669409783843339f3): Fixed code review comments.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/fc259ce518132787efff1cf669409783843339f3..6bf2db6e17030bfe16049fe1144e30c97d12315d): Moved ConfigMap namespace and name back to loader.go.

orelmisan commented on kiagnose/internal/client/client.go:
```diff
@@ -0,0 +1,39 @@
+	"k8s.io/client-go/rest"
+)
+
+func GetK8sClientset() (*kubernetes.Clientset, error) {
```
Done.

EdDev reviewed (approved):
Thank you!

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/23ab02e01cef696d6b1c3d0c603e17751abe48c0..6db629219d4484072c5541b1c9726d49edcc5e6b): Rebase and integration with PR #9.

ormergi commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,165 @@
+	ConfigMapNameEnvVarName      = "CONFIGMAP_NAME"
+)
+
+type Loader struct {
```
It would make Loader{} easier to change (e.g moving to CRD) if it would hold an instance of a parser,
it could also make it easier to test as you could replace it with a stub on tests.

orelmisan replied:
The tests in loader_test.go are also covering 100% of the parser code.

EdDev replied:
In this case, I think it is actually very easy to keep the parser internal and not expose it as an API.
By doing so, the tests are not aware it even exists and therefore should not be affected if it changes (or replaced).

There are cases where this is not possible (e.g. the k8s client) due to external dependencies, but here it is.
It is also easy to check all permutations through the `Loader` API, at least with the current input/output logic options.

ormergi replied:
@EdDev understood, thanks for explaining!

ormergi commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,305 @@
+	}
+}
+
+func TestLoaderLoadShouldFail(t *testing.T) {
```
Please add tests to cover client errors e.g: when it fails to fetch Role's ClusterRole's or ConfigMap's
similar to what we do at `checkup` tests.

orelmisan replied:
This is covered by three tests:
1. [Failure to get a ConfigMap](https://github.com/kiagnose/kiagnose/pull/15/files#diff-9c52ee6d71999c83d9e358fc8c604002291553b5351a4c710ac89fdc1b662206R157).
2. [Failure to get a ClusterRole](https://github.com/kiagnose/kiagnose/pull/15/files#diff-9c52ee6d71999c83d9e358fc8c604002291553b5351a4c710ac89fdc1b662206R192).
3. [Failure to get a Role](https://github.com/kiagnose/kiagnose/pull/15/files#diff-9c52ee6d71999c83d9e358fc8c604002291553b5351a4c710ac89fdc1b662206R197).

EdDev replied:
The client errors seem to be covered by just making sure the fetched object is not in the fake client.
It was illusive to me too at first.

ormergi commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,305 @@
+		},
+	}
+
+	for _, testCase := range failureTestCases {
+		t.Run(testCase.description, func(t *testing.T) {
+			fakeClient := newFakeClientWithObjects(configMapNamespace, configMapName, testCase.configMapData, nil, nil)
+
+			loader, err := config.NewLoader(fakeClient, goodEnv)
+			assert.NoError(t, err)
+
+			_, err = loader.Load()
+			assert.ErrorContains(t, err, testCase.expectedError)
+		})
+	}
```
I think those tests should have their own test suite at `cmparser` package, 
it would make it easier to test Loader{} as you could replace it the parser with a stub.
Will you give it a try?

orelmisan replied:
The tests in loader_test.go are also covering 100% of the parser code.

ormergi commented on kiagnose/mainflow.go:
```diff
@@ -20,17 +20,32 @@
 )
 
-func Run() error {
+func Run(env map[string]string) error {
```
I think mainflow.Run() should read env by it self instead of receiving it as parameter,
it should at least validate ConfigMap name and namespace env vars are valid as it is critical 
requirement for it to work.
It is not the scope of the "config loader" to validate it.

orelmisan replied:
The validation was moved to Loader.Load().

orelmisan commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,165 @@
+	}
+
+	return configMap.Data, nil
+}
```
There is now a configmap package containing both functions.

ormergi replied:
Great, thanks!

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/6db629219d4484072c5541b1c9726d49edcc5e6b..088d2010d0fcb2ddb72db853ea5c54b7e2e03b4f): Added a configmap package to unify actions against a ConfigMap object.

orelmisan commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,165 @@
+
+		role, err := l.client.RbacV1().Roles(fullRoleName.Namespace).Get(context.Background(), fullRoleName.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
```
Done.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/088d2010d0fcb2ddb72db853ea5c54b7e2e03b4f..d24bc743c40da32d9b9058d1ef1e5188b6b18aed): Put the logic of `splitRoleName()` inline.

orelmisan commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,305 @@
+			actualConfig, err := loader.Load()
+			assert.NoError(t, err)
+
+			sort.Slice(actualConfig.EnvVars, func(i, j int) bool {
+				return actualConfig.EnvVars[i].Name < actualConfig.EnvVars[j].Name
+			})
+
+			assert.Equal(t, testCase.expectedConfig, actualConfig)
```
IIUC assert.ElementsMatch() works on slices.
Here we have a slice inside a struct so it won't help in this case because we are comparing structs.

orelmisan commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,305 @@
+}
+
+func TestLoaderLoadShouldFail(t *testing.T) {
+	t.Run("when ConfigMap doesn't exist", testLoaderFailureWhenConfigMapIsMissing)
```
Done.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/d24bc743c40da32d9b9058d1ef1e5188b6b18aed..230c3553275929c26f313afcec5ba8b57c7f281f): Inlined testLoaderFailureWhenConfigMapIsMissing()

orelmisan commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,165 @@
+	configMapName      string
+}
+
+func NewLoader(client kubernetes.Interface, env map[string]string) (*Loader, error) {
```
As we have discussed offline, the environment variables validation logic was moved to Loader.Load().
NewLoader() doesn't return an error any more.

ormergi replied:
Great, thanks it looks much better.

orelmisan commented:
Change: Fixed commit messages.

EdDev commented on kiagnose/internal/configmap/configmap.go:
```diff
@@ -25,9 +25,19 @@ import (
 	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
 )
 
+func GetData(client kubernetes.Interface, namespace, name string) (map[string]string, error) {
```
The commit says it moves the the package up, but in fact you also introduce a new method.

I would just move it, reasoning that it also is to be shared with the config package.
Then, when you actually need that method, to introduce it.

BTW, when you add it, it will fit nicer below the `Create` (just a nit).

orelmisan replied:
Done.

EdDev reviewed (commented):
First commit

EdDev commented on kiagnose/internal/rbac/rbac.go:
```diff
@@ -32,10 +32,29 @@ import (
 	rbacv1client "k8s.io/client-go/kubernetes/typed/rbac/v1"
 )
 
+func GetClusterRole(client kubernetes.Interface, name string) (*rbacv1.ClusterRole, error) {
```
Same here, better introduce it when you actually use it.

orelmisan replied:
Done.

EdDev reviewed (commented):
Second commit.

EdDev commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,141 @@
+		return nil, fmt.Errorf(envVarErrFmtString, ConfigMapNameEnvVarName)
+	}
+
+	var err error
```
I think you need this only further below. Try to place it close to it being needed.

EdDev commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,141 @@
+	}
+
+	parser := newConfigMapParser(rawData)
+	err = parser.Parse()
+	if err != nil {
```
Can be inlined

orelmisan replied:
Without this, one linter fails, without the previous one, a different linter fails.

EdDev replied:
Why would `if err := parser.Parse(); err != nil {...}` fail?

orelmisan replied:
```
if err := parser.Parse(); err != nil {
	return nil, err
}
```

```
./automation/make.sh 
kiagnose/internal/config/loader.go:68:5: shadow: declaration of "err" shadows declaration at line 61 (govet)
        if err := parser.Parse(); err != nil {
           ^
```

EdDev replied:
Pfff... just a matter of naming (and line 61 should have been removed).
Not going to block on this nit.

EdDev reviewed (commented):
commit 3

orelmisan commented:
Change: Changed commit messages.

EdDev reviewed (approved):
Thank you!

ormergi commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,290 @@
+	configMapData map[string]string, clusterRoles []*rbacv1.ClusterRole, roles []*rbacv1.Role) *fake.Clientset {
+	client := fake.NewSimpleClientset(newConfigMap(namespace, name, configMapData))
+
+	for i := range clusterRoles {
+		_, err := client.RbacV1().ClusterRoles().Create(context.Background(), clusterRoles[i], metav1.CreateOptions{})
+		if err != nil {
+			panic("failed to create ClusterRole")
+		}
+	}
+
+	for i := range clusterRoles {
+		_, err := client.RbacV1().Roles(roles[i].Namespace).Create(context.Background(), roles[i], metav1.CreateOptions{})
+		if err != nil {
+			panic("failed to create Role")
+		}
+	}
```
I think this can be simplified by passing `[]ClusterRoles` and `[]Roles` directly to `fake.NewSimpleClientset()`.

orelmisan replied:
I've had problems making it compile.

ormergi replied:
Sorry for the noise, I checked it again and it works when passing *RoleList and *ClusterRoleList instead of []*Role and []*ClusterRoleList.

ormergi commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,290 @@
+			expectedError: "roles.rbac.authorization.k8s.io",
+		},
+		{
+			description:   "when Role name is illegal",
```
Do you think it make sense to cover more cases of `RolesKey`, such as `/myrole` `myns/` `ns//role` `""`?

What you think about adding similar test for `ClusterRoleKey` does it make sense?

orelmisan replied:
There is a full ticket regarding configuration validation.
IMO this should be handled there.

EdDev replied:
The code was introduced in this change, so tests should have been covering various scenarios of it here.
Said that, I will take the chance and take this change in even without this part.
Consider improving this in a following change.

ormergi commented on kiagnose/mainflow.go:
```diff
@@ -41,7 +39,6 @@ func Run(env map[string]string) error {
 	}
 
-	log.Printf("Checkup configuration: %+v", checkupConfig)
-	l := launcher.New(checkup.New(c, "", 0, nil, nil, nil), reporter.New())
```
Looks like a leftover

EdDev replied:
?

orelmisan replied:
This is a code that was deleted.
What is wrong with it?

ormergi replied:
Correct me if I am wrong but this change was introduced on commit #4 and and deleted at commit #5.

ormergi replied:
Correct me if I am wrong but this change (line 44) was introduced on commit #4 and and deleted at commit #5.

orelmisan replied:
Commit 4 is to introduce the Config loading to mainflow.
Commit 5 is to integrate the config loading with the checkup.

Is there something do you think we should do here?

ormergi replied:
Remove this line https://github.com/kiagnose/kiagnose/pull/15/commits/79420429daaf952650e57f41cb4f37c7a47ace58#diff-4ec73023f96b17dcaab6fb900294476b4a985df1fac89d1545944edc5fc6e163R44

orelmisan replied:
It won't compile without it.
Unless we do:
```go
_, err := configLoader.Load()
```

orelmisan replied:
@ormergi Done.

ormergi commented:
@orelmisan 
Thanks for the changes it look much better!

I have some more comments, most of them are suggestions to simplify the code even more,
Please take a look and see if it make sense to do.

EdDev reviewed (commented):
I answered here in an attempt to separate the nits from the ones which merit attention once detected.
Looking forward to see this finalized.

orelmisan commented on kiagnose/internal/config/cmparser.go:
```diff
@@ -0,0 +1,150 @@
+	roleNames        []string
+}
+
+func newConfigMapParser(configMapRawData map[string]string) *configMapParser {
+	return &configMapParser{
+		configMapRawData: configMapRawData,
+		params:           map[string]string{},
+	}
+}
+
+func (cmp *configMapParser) Parse() error {
```
The parser internal state is different than the Config struct.
The image and timeout fields are the same but:
1. The parser holds raw map of strings for params, while Config holds []EnvVar.
2. The parser holds names of ClusterRole and Role objects, while Config holds a slices of ClusterRole and Role.

ormergi replied:
> Considering that the parsing includes multiple functions, each requiring to pass the raw data, the usage of receivers seems reasonable to me.

Just to be clear, changing Parse() to static also means getting rid of the helpers receivers so each will be static as well.

I am not insisting on this change.

ormergi replied:
@orelmisan I didnt meant to use `configMapParser` as the result of Parse() not `Config` struct.

orelmisan replied:
IIUC you want the configMapParser (lets say ParseResults) as the result of Parse().
We can do it in a future PR.

ormergi replied:
Fair enough

orelmisan commented on kiagnose/internal/config/cmparser.go:
```diff
@@ -0,0 +1,150 @@
+)
+
+var (
+	ErrConfigMapDataIsNil    = errors.New("configMap Data field is nil")
+	ErrImageFieldIsMissing   = errors.New("image field is missing")
+	ErrTimeoutFieldIsMissing = errors.New("timeout field is missing")
+	ErrTimeoutFieldIsIllegal = errors.New("timeout field is illegal")
+	ErrRoleNameIsIllegal     = errors.New("illegal Role name")
```
This allows us to publish (some) of the errors that this package produces.

ormergi replied:
I am not talking about formatting strings, just extracting the message to const and create the error where needed:
```
const errNotValid = "not valid"
...
func foo () err {
    return errors.New("errNotValid")
}
```
Currently we do not provide errors as part of the API, isnt it?

ormergi replied:
@EdDev  I am not talking about formatting strings, just extracting the message to const and create the error where needed:
```
const errNotValid = "not valid"
...
func foo () err {
    return errors.New("errNotValid")
}
```
@orelmisan the errors are only used at tests, and they can use the error message instead.
I think this package should not expose the errors until there is a need for it.

ormergi replied:
@orelmisan @EdDev thanks for explaining.

Looking at this again, I was confused since on tests the assertions is done on the error message content and not on its type, I guess we can improve them later on.

orelmisan commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,141 @@
+}
+
+func (l *Loader) Load() (*Config, error) {
+	const envVarErrFmtString = "missing required environment variable %q"
```
Done.

orelmisan commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,141 @@
+		return nil, err
+	}
+
+	checkupConfig := &Config{
+		Image:   parser.Image(),
+		Timeout: parser.Timeout(),
+		EnvVars: paramsToEnvVars(parser.Params()),
+	}
```
Done.

ormergi replied:
Sorry I wasn't clear about saying this is a nit.

orelmisan commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,141 @@
+	return checkupConfig, nil
+}
+
+func (l *Loader) getClusterRoles(clusterRoleNames []string) ([]*rbacv1.ClusterRole, error) {
```
Done.

orelmisan commented on kiagnose/internal/config/loader.go:
```diff
@@ -0,0 +1,141 @@
+	return clusterRoles, nil
+}
+
+func (l *Loader) getRoles(roleNames []string) ([]*rbacv1.Role, error) {
```
Done.

ormergi commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,290 @@
+		t.Run(testCase.description, func(t *testing.T) {
+			loader := config.NewLoader(nil, testCase.envVars)
+			_, err := loader.Load()
+			assert.ErrorContains(t, err, testCase.expectedErrorContains)
```
Errors may change over time meaning it will be necessary to change the tests accordingly, 
I think we should avoid assenting on error message content when possible.

Using `assert.Error()` will simplify the tests since there will be no need to pass `expectedErrorContains`
on each test case.

orelmisan replied:
Here we want to differentiate between different errors, not just to assert that an error had occurred.

orelmisan replied:
IMO we should check that the error we expect is the actual error we are getting.

EdDev replied:
The cleanest solution would have been to assert against an error type or instance.
You can introduce improvements to this in following changes if you like, at this stage I do not consider this a blocker.

orelmisan commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,290 @@
+
+	for _, testCase := range failureTestCases {
+		t.Run(testCase.description, func(t *testing.T) {
+			fakeClient := newFakeClientWithObjects(configMapNamespace, configMapName, testCase.configMapData, nil, nil)
```
Done.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/fcff5c7929d9f62e7d54623a7d0479367d6560bd..7197d634f0a7cf4d9a7e338ff614d2426ca700b4): Answered code review comments.

orelmisan commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,290 @@
+
+	for _, testCase := range testCases {
+		t.Run(testCase.description, func(t *testing.T) {
+			fakeClient := newFakeClientWithObjects(configMapNamespace, configMapName, testCase.configMapData, testCase.clusterRoles, testCase.roles)
```
I've had problems making this compile.

EdDev replied:
I agree this is an improvement, but I will take this change in even without it.
Feel free to post this change in a following PR.

orelmisan commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,290 @@
+
+func expectedClusterRoles() []*rbacv1.ClusterRole {
+	return []*rbacv1.ClusterRole{
+		{
+			TypeMeta:   metav1.TypeMeta{Kind: "ClusterRole", APIVersion: rbacv1.GroupName},
+			ObjectMeta: metav1.ObjectMeta{Name: "cluster_role1"},
+		},
+		{
+			TypeMeta:   metav1.TypeMeta{Kind: "ClusterRole", APIVersion: rbacv1.GroupName},
+			ObjectMeta: metav1.ObjectMeta{Name: "cluster_role2"},
+		},
```
Can you please elaborate?

ormergi replied:
```
{TypeMeta: metav1.TypeMeta{Kind: "ClusterRole", APIVersion: rbacv1.GroupName},ObjectMeta: metav1.ObjectMeta{Name: "cluster_role1"}},
```

orelmisan replied:
IMO it makes it less readable.

orelmisan commented on kiagnose/internal/config/loader_test.go:
```diff
@@ -0,0 +1,290 @@
+
+func expectedRoles() []*rbacv1.Role {
+	return []*rbacv1.Role{
+		{
+			TypeMeta:   metav1.TypeMeta{Kind: "Role", APIVersion: rbacv1.GroupName},
+			ObjectMeta: metav1.ObjectMeta{Name: "role1", Namespace: "default"},
+		},
+		{
+			TypeMeta:   metav1.TypeMeta{Kind: "Role", APIVersion: rbacv1.GroupName},
+			ObjectMeta: metav1.ObjectMeta{Name: "role2", Namespace: "default"},
+		},
+	}
```
Can you please elaborate?

ormergi replied:
```
{TypeMeta: metav1.TypeMeta{Kind: "Role", APIVersion: rbacv1.GroupName}, ObjectMeta: metav1.ObjectMeta{Name: "role2", Namespace: "default"}}
```

orelmisan replied:
IMO it makes it less readable.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/7197d634f0a7cf4d9a7e338ff614d2426ca700b4..de035b2fcca73c3908c64e56d18e50ab190f19e6): Fixed a small bug in `newFakeClientWithObjects()`

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/de035b2fcca73c3908c64e56d18e50ab190f19e6..4cd6195c2b5ec591ef11d3f2fad0210d3e9642ea): Fixed a review comment.

ormergi reviewed (approved):
Thanks for the changes, good work!

There are these two comments https://github.com/kiagnose/kiagnose/pull/15#discussion_r852673792
https://github.com/kiagnose/kiagnose/pull/15#discussion_r852173990, if you think it will take more work then it seems we can do it on a follow up PR.

EdDev reviewed (approved):
Thank you!

I have considered the still open issues from the reviews and I think it is safe to merge this as is.
Thank you @ormergi for the thorough review.
